ngraph/test/util/engine/ie_engines.cpp

```diff
@@ -0,0 +1,230 @@
+            return compare_blobs<int8_t>(computed, expected, tolerance_bits);
+            break;
+        case InferenceEngine::Precision::I16:
+            return compare_blobs<int16_t>(computed, expected, tolerance_bits);
+            break;
+        case InferenceEngine::Precision::I32:
+            return compare_blobs<int32_t>(computed, expected, tolerance_bits);
+            break;
+        case InferenceEngine::Precision::I64:
+            return compare_blobs<int64_t>(computed, expected, tolerance_bits);
+            break;
+        case InferenceEngine::Precision::U8:
+            return compare_blobs<uint8_t>(computed, expected, tolerance_bits);
+            break;
+        case InferenceEngine::Precision::U16:
+            return compare_blobs<uint16_t>(computed, expected, tolerance_bits);
+            break;
+        case InferenceEngine::Precision::U64:
+            return compare_blobs<uint64_t>(computed, expected, tolerance_bits);
+            break;
+        case InferenceEngine::Precision::BOOL:
+            return compare_blobs<uint8_t>(computed, expected, tolerance_bits);
+            break;
+        default: THROW_IE_EXCEPTION << "Not implemented yet";
+        }
+    }
+};
+
+test::IE_Engine::IE_Engine(const std::shared_ptr<Function> function, const char* device)
+    : m_function{function}
+{
+    upgrade_and_validate_function(m_function);
+    const auto cnn_network = InferenceEngine::CNNNetwork(m_function);
+    m_network_inputs = cnn_network.getInputsInfo();
+    m_network_outputs = cnn_network.getOutputsInfo();
+
+    InferenceEngine::Core ie;
+    auto exe_network = ie.LoadNetwork(cnn_network, device);
+    m_inference_req = exe_network.CreateInferRequest();
+}
+
+void test::IE_Engine::infer()
+{
+    if (m_network_inputs.size() != m_allocated_inputs)
+    {
+        THROW_IE_EXCEPTION << "The tested graph has " << m_network_inputs.size() << " inputs, but "
+                           << m_allocated_inputs << " were passed.";
+    }
+    else
+    {
+        m_inference_req.Infer();
+    }
+}
+
+testing::AssertionResult test::IE_Engine::compare_results(const size_t tolerance_bits)
+{
+    auto comparison_result = testing::AssertionSuccess();
+
+    for (const auto& output : m_network_outputs)
+    {
+        InferenceEngine::MemoryBlob::CPtr computed_output_blob =
+            InferenceEngine::as<InferenceEngine::MemoryBlob>(m_inference_req.GetBlob(output.first));
+
+        const auto& expected_output_blob = m_expected_outputs[output.first];
+
+        comparison_result =
+            compare_blobs(computed_output_blob, expected_output_blob, tolerance_bits);
+
+        if (comparison_result == testing::AssertionFailure())
+        {
+            break;
+        }
+    }
+
+    return comparison_result;
+}
+
+std::shared_ptr<Function>
+    test::IE_Engine::upgrade_and_validate_function(const std::shared_ptr<Function> function) const
+{
+    pass::Manager passes;
+    passes.register_pass<pass::Opset1Upgrade>();
```

Comment:
Since filename changed I have to past it here again:
I'm not sure if test engine should upgrade a given function opset.
It can shadow a problem when ONNX importer or any other component will create a opset0 operation by mistake.

Reply:
This is also covered in a separate ticket. Check out 33971.